File: packages/core/src/trace.ts
export interface TraceparentInput {
  traceId?: string;
  spanId?: string;
  traceFlags?: number | string;
}

const traceIdPattern = /^[0-9a-f]{32}$/;
const spanIdPattern = /^[0-9a-f]{16}$/;
const traceFlagsPattern = /^[0-9a-f]{2}$/;

export function createTraceparent({
  traceId,
  spanId,
  traceFlags = "01",
}: TraceparentInput): string {
  const normalizedTraceId = normalizeTraceId(traceId);
  const normalizedSpanId = normalizeSpanId(spanId);
  const normalizedTraceFlags = normalizeTraceFlags(traceFlags);
  if (!normalizedTraceId || !normalizedSpanId || !normalizedTraceFlags) {
    return "";
  }
  return `00-${normalizedTraceId}-${normalizedSpanId}-${normalizedTraceFlags}`;
}

// A valid id matches its hex-length pattern and is not the all-zero sentinel
// (which the regex admits but OTel uses to mean "no id").
function normalizeId(id: string | undefined, pattern: RegExp, zero: string): string {
  const normalized = id?.trim().toLowerCase() ?? "";
  if (!pattern.test(normalized) || normalized === zero) return "";
  return normalized;
}

function normalizeTraceId(traceId: string | undefined): string {
  return normalizeId(traceId, traceIdPattern, "00000000000000000000000000000000");
}

function normalizeSpanId(spanId: string | undefined): string {
  return normalizeId(spanId, spanIdPattern, "0000000000000000");
}

function normalizeTraceFlags(traceFlags: number | string): string {
  if (typeof traceFlags === "number") {
    if (!Number.isInteger(traceFlags) || traceFlags < 0 || traceFlags > 255) {
      return "";
    }
    return traceFlags.toString(16).padStart(2, "0");
  }
  const normalized = traceFlags.trim().toLowerCase();
  return traceFlagsPattern.test(normalized) ? normalized : "";
}

File: packages/core-py/src/astropods_adapter_core/trace.py
from __future__ import annotations

import re
from typing import Union

_TRACE_ID_PATTERN = re.compile(r"^[0-9a-f]{32}$")
_SPAN_ID_PATTERN = re.compile(r"^[0-9a-f]{16}$")
_TRACE_FLAGS_PATTERN = re.compile(r"^[0-9a-f]{2}$")


def create_traceparent(
    *,
    trace_id: str | None,
    span_id: str | None,
    trace_flags: Union[int, str] = "01",
) -> str:
    normalized_trace_id = _normalize_trace_id(trace_id)
    normalized_span_id = _normalize_span_id(span_id)
    normalized_trace_flags = _normalize_trace_flags(trace_flags)
    if not normalized_trace_id or not normalized_span_id or not normalized_trace_flags:
        return ""
    return f"00-{normalized_trace_id}-{normalized_span_id}-{normalized_trace_flags}"


# A valid id matches its hex-length pattern and is not the all-zero sentinel
# (which the regex admits but OTel uses to mean "no id").
def _normalize_id(id_: str | None, pattern: "re.Pattern[str]", zero: str) -> str:
    normalized = (id_ or "").strip().lower()
    if not pattern.match(normalized) or normalized == zero:
        return ""
    return normalized


def _normalize_trace_id(trace_id: str | None) -> str:
    return _normalize_id(trace_id, _TRACE_ID_PATTERN, "00000000000000000000000000000000")


def _normalize_span_id(span_id: str | None) -> str:
    return _normalize_id(span_id, _SPAN_ID_PATTERN, "0000000000000000")


def _normalize_trace_flags(trace_flags: Union[int, str]) -> str:
    if isinstance(trace_flags, int):
        if trace_flags < 0 or trace_flags > 255:
            return ""
        return f"{trace_flags:02x}"
    normalized = trace_flags.strip().lower()
    if not _TRACE_FLAGS_PATTERN.match(normalized):
        return ""
    return normalized
